File: python/algo/202506/algo0818.py
from cmath import inf
from functools import cache
from heapq import heappop, heappush


class Solution:
    def racecar(self, target: int) -> int:
        # 自行解答 广度优先搜索 + 优先队列
        n = step = 1 
        while n <= target: n <<= 1

        dp = [inf] * n
        dp[0] = 0
        q = []
        while (1 << step) - 1 < n:
            heappush(q, (step, (1 << step)-1, 1))
            step += 1

        while q:
            val, pos, dir = heappop(q)
            if pos == target: return val
            if val > dp[pos]: continue
            else: dp[pos] = val
            # 反方向
            step = 1
            while 0 < pos + dir * -1 * ((1<<step) - 1) < n:
                nxtPos = pos + dir * -1 * ((1<<step) - 1)
                if dp[nxtPos] > val + step + 1:
                    heappush(q, (val + step + 1, nxtPos, dir*-1))
                step += 1
            # 同方向
            step = 1
            while 0 < pos + dir * ((1<<step) - 1) < n:
                nxtPos = pos + dir * ((1<<step) - 1)
                if dp[nxtPos] > val + step + 2:
                    heappush(q, (val + step + 2, nxtPos, dir))
                step += 1

        return -1


if __name__ == "__main__":
    sol = Solution()
    print(sol.racecar(3))

File: cplus/leetcode/202506/algo0818.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 -  一维 DP
class Solution {
public:
    int racecar(int target) {
        int n = 1, step = 1;
        while (n <= target) n <<= 1;
        vector<int> dp(n, INT_MAX);

        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<>> que;

        dp[0] = 0;
        while ((1<<step) - 1 < n) {
            que.emplace(step, (1<<step)-1, 1);
            step ++;
        }

        auto nxt = [](int pos, int dir, int step) -> int{
            return pos + dir * ((1<<step)-1);
        };

        auto check = [&] (int pos, int dir, int step) -> bool{
            int new_pos = nxt(pos, dir, step);
            return new_pos >0 && new_pos < n;
        };

        while (!que.empty()) {
            auto [val, pos, dir] = que.top(); que.pop();
            if (pos == target) return val;
            if (val > dp[pos]) continue; // 剪枝，否则超时
            else dp[pos] = val;          // 关键，出队列时，设置值
            // 同方向继续 AA， + 2 
            step = 1;
            while (check(pos, dir, step)) {
                int nxtPos = nxt(pos, dir, step);
                if (dp[nxtPos] > val + 2 + step) {
                    que.emplace(val + 2 + step, nxtPos, dir);
                }
                step += 1;
            }
            // 反方向继续 A， + 1 
            step = 1;
            while (check(pos, dir*-1, step)) {
                int nxtPos = nxt(pos, dir*-1, step);
                if (dp[nxtPos] > val + 1 + step) {
                    que.emplace(val + 1 + step, nxtPos, dir*-1);
                }
                step += 1;
            }
        }

        return -1;     
    }
};
